### PR TITLE
Add typed IDs

### PR DESCRIPTION
**Pull request by tadeokondrak**

### CODE REVIEW

**Comment by diamondburned** [comment]
Can you change the target branch to `v1`?

File: discord/snowflake.go [hunk 1]
```diff
@@ -73,6 +73,11 @@ func (s Snowflake) Valid() bool {
 	return int64(s) > 0
 }
 
+// IsNull returns whether or not the snowflake is null.
+func (s Snowflake) IsNull() bool {
+	return s == NullSnowflake
+}
+
```
Comment (tadeokondrak) [inline discord/snowflake.go:80]: This is new

File: discord/user.go [hunk 2]
```diff
@@ -93,9 +93,9 @@ const (
 )
 
 type Connection struct {
-	ID   Snowflake `json:"id"`
-	Name string    `json:"name"`
-	Type Service   `json:"type"`
+	ID   string  `json:"id"`
```
Comment (tadeokondrak) [inline discord/user.go:96]: ID was changed

File: gateway/commands.go [hunk 3]
```diff
@@ -144,12 +144,12 @@ func (g *Gateway) UpdateStatusCtx(ctx context.Context, data UpdateStatusData) er
...
-	Typing     bool              `json:"typing"`
-	Activities bool              `json:"activities"`
-	GuildID    discord.Snowflake `json:"guild_id"`
+	Typing     bool            `json:"typing"`
+	Activities bool            `json:"activities"`
+	GuildID    discord.GuildID `json:"guild_id"`
 
 	// Channels is not documented. It's used to fetch the right members sidebar.
-	Channels map[discord.Snowflake][][2]int `json:"channels"`
+	Channels map[discord.ChannelID][][2]int `json:"channels"`
```
Comment (tadeokondrak) [inline gateway/commands.go:152]: I'm pretty sure this is correct but not 100%

File: gateway/events.go [hunk 4]
```diff
@@ -122,18 +122,18 @@ type (
...
+		OnlineCount uint64          `json:"online_count"`
 
 		// Groups is all the visible role sections.
 		Groups []GuildMemberListGroup `json:"groups"`
 
 		Ops []GuildMemberListOp `json:"ops"`
 	}
 	GuildMemberListGroup struct {
-		ID    string `json:"id"` // either discord.Snowflake Role IDs or "online"
+		ID    string `json:"id"` // either discord.RoleID or "online"
```
Comment (tadeokondrak) [inline gateway/events.go:136]: Original comment had plural IDs, is that correct?

File: gateway/ready.go [hunk 5]
```diff
@@ -102,13 +102,13 @@ type SettingsChannelOverride struct {
...
 }
 
 // GuildFolder holds a single folder that you see in the left guild panel.
 type GuildFolder struct {
-	Name     string              `json:"name"`
-	ID       discord.Snowflake   `json:"id"`
-	GuildIDs []discord.Snowflake `json:"guild_ids"`
-	Color    discord.Color       `json:"color"`
+	Name     string            `json:"name"`
+	ID       string            `json:"id"`
```
Comment (tadeokondrak) [inline gateway/ready.go:111]: ID was changed

File: bot/extras/infer/infer.go [hunk 6]
```diff
@@ -13,21 +13,21 @@ import (
...
 }
 
 // UserID looks for fields with name UserID, User, or in some special cases, ID.
-func UserID(event interface{}) discord.Snowflake {
+func UserID(event interface{}) discord.UserID {
 	// This may have a very fatal bug of accidentally mistaking another User's
 	// ID. It also probably wouldn't work with things like RecipientID.
-	return reflectID(reflect.ValueOf(event), "User")
+	return discord.UserID(reflectID(reflect.ValueOf(event), "User"))
 }
```
Comment (tadeokondrak) [inline bot/extras/infer/infer.go:31]: Does this file need to be changed with typed IDs?

File: api/member.go [hunk 7]
```diff
@@ -247,7 +247,7 @@ func (c *Client) Prune(guildID discord.Snowflake, data PruneData) (uint, error)
 //
 // Requires KICK_MEMBERS permission.
 // Fires a Guild Member Remove Gateway event.
-func (c *Client) Kick(guildID, userID discord.Snowflake) error {
+func (c *Client) Kick(guildID, userID discord.UserID) error {
```
Comment (tadeokondrak) [inline api/member.go:250]:
```suggestion
func (c *Client) Kick(guildID discord.GuildID, userID discord.UserID) error {
```

File: bot/extras/infer/infer.go (hunk 6 above)
Comment (diamondburned) [reply bot/extras/infer/infer.go:31]: Yeah, this seems correct.

File: gateway/events.go (hunk 4 above)
Comment (diamondburned) [reply gateway/events.go:136]: Probably not. Singular seems to be valid.

File: gateway/commands.go (hunk 3 above)
Comment (diamondburned) [reply gateway/commands.go:152]: It is correct.

File: api/channel.go [hunk 8]
```diff
@@ -181,7 +181,7 @@ type EditChannelPermissionData struct {
 //
 // Requires the MANAGE_ROLES permission.
 func (c *Client) EditChannelPermission(
-	channelID, overwriteID discord.Snowflake, data EditChannelPermissionData) error {
+	channelID discord.ChannelID, overwriteID discord.Snowflake, data EditChannelPermissionData) error {
```
[the remark this answers is not in the document]
Comment (diamondburned) [reply api/channel.go:184]: This is documented, so it should be fine.

**Pull request merged**